crates/ra_vfs/src/watcher.rs

```diff
@@ -0,0 +1,96 @@
+use std::{
+    path::{Path, PathBuf},
+    sync::mpsc,
+    thread,
+    time::Duration,
+};
+
+use crate::io;
+use crossbeam_channel::Sender;
+use drop_bomb::DropBomb;
+use notify::{DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher as NotifyWatcher};
+
+pub struct Watcher {
+    watcher: RecommendedWatcher,
+    thread: thread::JoinHandle<()>,
+    bomb: DropBomb,
+}
+
+#[derive(Debug)]
+pub enum WatcherChange {
+    Create(PathBuf),
+    Write(PathBuf),
+    Remove(PathBuf),
+    Rescan,
+}
+
+fn send_change_events(
+    ev: DebouncedEvent,
+    sender: &Sender<io::Task>,
+) -> Result<(), Box<std::error::Error>> {
+    match ev {
+        DebouncedEvent::NoticeWrite(_)
+        | DebouncedEvent::NoticeRemove(_)
+        | DebouncedEvent::Chmod(_) => {
+            // ignore
+        }
+        DebouncedEvent::Rescan => {
+            sender.send(io::Task::LoadChange(WatcherChange::Rescan))?;
+        }
+        DebouncedEvent::Create(path) => {
+            sender.send(io::Task::LoadChange(WatcherChange::Create(path)))?;
+        }
+        DebouncedEvent::Write(path) => {
+            sender.send(io::Task::LoadChange(WatcherChange::Write(path)))?;
+        }
+        DebouncedEvent::Remove(path) => {
+            sender.send(io::Task::LoadChange(WatcherChange::Remove(path)))?;
+        }
+        DebouncedEvent::Rename(src, dst) => {
+            sender.send(io::Task::LoadChange(WatcherChange::Remove(src)))?;
+            sender.send(io::Task::LoadChange(WatcherChange::Create(dst)))?;
+        }
+        DebouncedEvent::Error(err, path) => {
+            // TODO should we reload the file contents?
+            log::warn!("watcher error {}, {:?}", err, path);
+        }
+    }
+    Ok(())
+}
+
+impl Watcher {
+    pub(crate) fn start(
+        output_sender: Sender<io::Task>,
+    ) -> Result<Watcher, Box<std::error::Error>> {
+        let (input_sender, input_receiver) = mpsc::channel();
+        let watcher = notify::watcher(input_sender, Duration::from_millis(250))?;
+        let thread = thread::spawn(move || {
+            input_receiver
+                .into_iter()
+                // forward relevant events only
+                .try_for_each(|change| send_change_events(change, &output_sender))
+                .unwrap()
+        });
```

Comment:
So this burns a thread to just transform `mpsc` channel into crossbeam one? Seems suboptimal, but I doubt we can do anything better.

Reply:
It's also mapping from `notify`s `DebouncedEvent` to `Tasks`s. I think even if the watcher took a crossbeam channel we would still need a thread to do the mapping